larpandoracontent/LArHelpers/LArMCParticleHelper.cc

```diff
@@ -598,31 +653,50 @@ void LArMCParticleHelper::GetPfoMCParticleHitSharingMaps(const PfoContributionMa
 //------------------------------------------------------------------------------------------------------------------------------------------
 
 void LArMCParticleHelper::CollectReconstructable2DHits(const ParticleFlowObject *const pPfo, const MCContributionMapVector &selectedMCParticleToHitsMaps,
-    pandora::CaloHitList &reconstructableCaloHitList2D)
+    pandora::CaloHitList &reconstructableCaloHitList2D, bool foldBackHierarchy)
 {
-    // Collect all 2D calo hits in pfo hierarchy
+
     PfoList pfoList;
-    LArPfoHelper::GetAllDownstreamPfos(pPfo, pfoList);
+
+    // If foldBackHierarchy collect all 2D calo hits in pfo hierarchy
+    // else collect hits directly belonging to pfo
+    if (foldBackHierarchy)
+    {
+        LArPfoHelper::GetAllDownstreamPfos(pPfo, pfoList);
+    }
+    else
+    {
+        pfoList.push_back(pPfo);
+    }
 
     LArMCParticleHelper::CollectReconstructable2DHits(pfoList, selectedMCParticleToHitsMaps, reconstructableCaloHitList2D);
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
 void LArMCParticleHelper::CollectReconstructableTestBeamHierarchy2DHits(const ParticleFlowObject *const pPfo, const MCContributionMapVector &selectedMCParticleToHitsMaps,
-    pandora::CaloHitList &reconstructableCaloHitList2D)
+    pandora::CaloHitList &reconstructableCaloHitList2D, bool foldBackHierarchy)
 {
-    // Collect all 2D calo hits in pfo hierarchy
+
     PfoList pfoList;
+    // If foldBackHierarchy collect all 2D calo hits in pfo hierarchy
+    // else collect hits directly belonging to pfo
 
-    // ATTN: Only collect downstream pfos for daughter test beam particles & cosmics
-    if (pPfo->GetParentPfoList().empty() && LArPfoHelper::IsTestBeam(pPfo))
+    if (foldBackHierarchy)
     {
-        pfoList.push_back(pPfo);
+        // ATTN: Only collect downstream pfos for daughter test beam particles & cosmics
+        if (pPfo->GetParentPfoList().empty() && LArPfoHelper::IsTestBeam(pPfo))
```

Comment:
For info: This is fiddly logic (existed before this PR), basically trying to handle the incoming test beam particle in a similar manner to a neutrino case. i.e. it's the chains of particles emerging from the interaction vertex that we want to collect. If we collect starting from the test beam particle itself, we'll only ever get one hierarchy.

Reply:
thank you for the info